### PR TITLE
Integrate tfds

### PR DESCRIPTION
**Pull request by ng390**
Armory integration script for existing public TFDS datasets. Runs as a script on armory container -- initial testing was with pytorch:0.13.1 dev image.

### CODE REVIEW

**Comment by ng390** [comment]
Marking as WIP because needs documentation link.

File: docs/integrate_tensorflow_datasets.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Instructions to Integrate TFDS Datasets
+
+1. Get the name, version number of the Tensorflow Dataset, and optionally the config: "name[/config]:version_number", where the brackets denote optional text.
+2. Set the environmental variables ARMORY_PRIVATE_S3_ID and ARMORY_PRIVATE_S3_KEY to the appropriate keys with write access to the Armory S3 bucket. 
+3. From a locally cloned version of armory, on a new branch, run:
+```
+python -m armory exec pytorch -- python integrate_tfds.py name[/config]:version
```
Comment (lcadalzo) [inline docs/integrate_tensorflow_datasets.md:7]: why pytorch?

File: docs/integrate_tensorflow_datasets.md [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# Instructions to Integrate TFDS Datasets
```
Comment (lcadalzo) [inline docs/integrate_tensorflow_datasets.md:1]: a minor note on the `.md` file: when viewing in the browser, items 4-10 all appear in one big paragraph as opposed to a list

File: docs/integrate_tensorflow_datasets.md [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+1. Get the name, version number of the Tensorflow Dataset, and optionally the config: "name[/config]:version_number", where the brackets denote optional text.
+2. Set the environmental variables ARMORY_PRIVATE_S3_ID and ARMORY_PRIVATE_S3_KEY to the appropriate keys with write access to the Armory S3 bucket. 
+3. From a locally cloned version of armory, on a new branch, run:
+```
+python -m armory exec pytorch -- python integrate_tfds.py name[/config]:version
+```
+where the brackets denote optional text.
+4. The script will download and process the TFDS dataset, generate TF Records files, create a tarball, and upload the tarball to S3. It also will create a S3 checksum file in ```armory/data/cached_s3_checksums/{name}.txt```
+5. Run ```git status``` to confirm the S3 checksum file was generated and to see the path of the template file.
+6. Manually put the template code from ```TEMPLATE_{name}.txt``` in ```armory/data/datasets.py```. Create a context object that contains metadata and a preprocessing function that does appropriate integrity checks/input normalizing.
```
Comment (lcadalzo) [inline docs/integrate_tensorflow_datasets.md:12]: I would modify the instructions regarding creating context objects/preprocessing functions (and `SUPPORTED_DATASETS`) to point to examples of this in the code, for clarity's sake

Comment (lcadalzo) [reply docs/integrate_tensorflow_datasets.md:12]: It would make sense to mention that we expect preprocessing functions to map inputs to canonical form. A user may then wonder what's meant by "canonical". Perhaps this is something we could improve our documentation of, or add a test for each modality to ensure that new datasets users add are in canonical form

**Comment by lcadalzo** [comment]
A couple comments from a code organization standpoint:

- I think the `armory/data/` directory may be a more fitting home for `integrate_tfds.py` than the base directory 
- should the function in `upload_file.py` go in `armory/data/utils.py` instead? If not, should the filename be renamed to `upload_file_to_s3.py` to be more concrete and/or moved to `armory/data/` directory?

**Comment by lcadalzo** [comment]
Separate comment regarding `amory/utils/progress_percentage.py`: is there a reason this file is needed when `armory/data/progress_percentage.py` already exists? If yes, should the functionality be moved to `armory/data/` directory, perhaps added to the same  `armory/data/progress_percentage.py` file?

**Comment by lcadalzo** [comment]
One thing the boilerplate code doesn't appear to account for (which may be intentional?), which will vary depending on the dataset, is the `as_supervised` and `supervised_xy_keys` kwargs passed to `_generator_from_tfds()`.

**Comment by ng390** [comment]
> One thing the boilerplate code doesn't appear to account for (which may be intentional?), which will vary depending on the dataset, is the `as_supervised` and `supervised_xy_keys` kwargs passed to `_generator_from_tfds()`.
> 
> Edit: as well as `variable_length`, `variable_y`.

Good catch - these were missing. I added all the missing fields for `_generator_from_tfds()`, except those that are passed only as kwargs elsewhere in the code (`class_ids` and `index`).

**Comment by ng390** [comment]
@lcadalzo This is ready for review.

File: docs/integrate_tensorflow_datasets.md [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+1. Get the name, version number of the Tensorflow Dataset, and optionally the config: "name[/config]:version_number", where the brackets denote optional text.
+2. Set the environmental variables ARMORY_PRIVATE_S3_ID and ARMORY_PRIVATE_S3_KEY to the appropriate keys with write access to the Armory S3 bucket. 
+3. From a locally cloned version of armory, on a new branch, run:
+```
+python -m armory exec pytorch -- python -m armory.data.integrate_tfds name[/config]:version
+```
+where the brackets denote optional text.
+
+4. The script will download and process the TFDS dataset, generate TF Records files, create a tarball, and upload the tarball to S3. It also will create a S3 checksum file in ```armory/data/cached_s3_checksums/{name}.txt```
```
Comment (lcadalzo) [inline docs/integrate_tensorflow_datasets.md:11]: Very minor, but step 4 isn't really a step (in the sense that it requires user action). I'd leave the text but attach it to step 3 which will make clear that it's describing what that command does

File: docs/integrate_tensorflow_datasets.md [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+5. Run ```git status``` to confirm the S3 checksum file was generated and to see the path of the template file.
+6. Manually put the template code from ```TEMPLATE_{name}.txt``` in ```armory/data/datasets.py```. Create a context object that contains metadata and a preprocessing function that does appropriate integrity checks/input normalizing. See for example the [canonical fixed-size image preprocessing function](https://github.com/twosixlabs/armory/blob/deb7a469bf4a7497d14fdd87eba6417b5e44589f/armory/data/datasets.py#L617-L631) which checks the shapes of an image, and renormalizes it to be in the appropriate range defined by the context object (typically 0.0-1.0) with a standard type. See the documentation on dataset [preprocessing](https://armory.readthedocs.io/en/latest/datasets/#preprocessing) for more details.
+7. [Optional] Add the dataset to the [SUPPORTED_DATASETS](https://github.com/twosixlabs/armory/blob/deb7a469bf4a7497d14fdd87eba6417b5e44589f/armory/data/datasets.py#L1498-L1511) dictionary by adding a key with the dataset's name and value of the dataset function from the template code.
```
Comment (lcadalzo) [inline docs/integrate_tensorflow_datasets.md:14]: Should this step be mandatory, assuming users want to use the dataset in an Armory scenario? Or perhaps should we add a sentence clarifying that this must be done in order to use the new dataset in an Armory scenario?

**Review by lcadalzo — APPROVED** [review]
LGTM
